lib/plugin/instance.rb

```diff
@@ -1018,6 +1018,34 @@ def add_topic_static_page(page, options = {}, &blk)
     StaticController::CUSTOM_PAGES[page] = blk ? { topic_id: blk } : options
   end
 
+  # Lets plugin define custom unsubscribe keys,
+  # set custom instance variables on the controller's unsubscribe action,
+  # and describe what unsubscribing for that key does.
+  #
+  # Example extracted from discourse-chat:
+  #
+  # controller_unsubscribe_options = proc do |key|
+  #   @chat_email_frequencies = UserOption.chat_email_frequencies.map do |freq, _|
+  #     [I18n.t("unsubscribe.chat_summary.#{freq}"), freq]
+  #   end
+  #
+  #   @current_chat_email_frequency = key.user.user_option.chat_email_frequency
+  # end
+  #
+  # unsubscribe_action = proc do |params|
+  #   if params[:chat_email_frequency]
+  #     user.user_option.update!(chat_email_frequency: params[:chat_email_frequency])
+  #  end
+  # end
+  #
+  # register_unsubscribe_key('chat_summary', controller_unsubscribe_options, unsubscribe_action)
+  #
+  # In conjunction with this, you have to define a connector under app/views/connectors/unsubscribe_options.
+  def register_unsubscribe_key(type, set_options_blk, perform_unsubscribe_blk)
+    add_to_class(EmailController, "set_unsubscribe_#{type}_options", &set_options_blk)
+    add_to_class(UnsubscribeKey, "#{type}_unsubscribe", &perform_unsubscribe_blk)
```

Comment:
Instead of having to define methods across both classes and having the unsubscribe logic spread between the controller and model, I think we can improve this by making plugins register a class instead. This is what I have in mind: 

```
class BaseEmailUnsubscriber 
  def initialize(unsubscribe_key)
    @unsubscribe_key = unsubscribe_key
  end

  def prepare_unsubscribe_options(controller)
    raise 'not implemented'
  end

  def unsubscribe(params)
    raise 'not implemented'
  end
end

class TopicEmailUnsubscriber < BaseEmailUnsubscriber
  def prepare_unsubscribe_options(controller)
    user = controller.instance_variable_get(:@user)
    attrs = {} 

    watching = TopicUser.notification_levels[:watching]
    topic = @unsubscribe_key.post&.topic || @unsubscribe_key.topic
    controller.instance_variable_set(:@topic, topic)
    controller.instance_variable_set(:@watching_topic, TopicUser.exists?(user_id: user.id, notification_level: watching, topic_id: topic.id))

    return if topic.category_id.blank?
    return if !CategoryUser.exists?(user_id: user.id, notification_level: CategoryUser.watching_levels, category_id: topic.category_id)

    controller.instance_variable_set(
      :@watched_count, 
      TopicUser.joins(:topic)
        .where(user: user, notification_level: watching)
        .where(topics: { category_id: topic.category_id })
        .count
    )
  end

  def unsubscribe(params)
    updated = false

    return updated if associated_topic.nil?

    if params[:unwatch_topic]
      TopicUser.where(topic_id: topic.id, user_id: user.id)
        .update_all(notification_level: TopicUser.notification_levels[:tracking])

      updated = true
    end

    if params[:unwatch_category] && associated_topic.category_id
      TopicUser.joins(:topic)
        .where(user: user, notification_level: TopicUser.notification_levels[:watching])
        .where(topics: { category_id: topic.category_id })
        .update_all(notification_level: TopicUser.notification_levels[:tracking])

      CategoryUser
        .where(user_id: user.id, category_id: topic.category_id, notification_level: CategoryUser.watching_levels)
        .destroy_all

      updated = true
    end

    if params[:mute_topic]
      TopicUser.where(topic_id: topic.id, user_id: user.id).update_all(notification_level: TopicUser.notification_levels[:muted])

      updated = true
    end

    updated   
  end
end
```

In the controller actions, the controller will lookup its registry and initialises the right class. 

```
{
  topic: TopicEmailUnsubscriber,
  digest: DigestEmailUnsubscriber,
  chat_summary: ChatSummaryEmailUnsubscriber
}
```

Reply:
Sounds like a good improvement 👍